remoting/subsystem/src/main/java/org/jboss/as/remoting/RemotingHttpUpgradeService.java

```diff
@@ -153,7 +157,34 @@ public synchronized void start(final StartContext context) throws StartException
 
             org.jboss.as.domain.management.SecurityRealm securityRealm = null;
             if (saslAuthenticationFactory == null && (securityRealm = injectedSecurityRealm.getOptionalValue()) != null) {
-                saslAuthenticationFactory = securityRealm.getSaslAuthenticationFactory();
+
+                final ClassLoader loader = WildFlySecurityManager.getClassLoaderPrivileged(ConnectorUtils.class);
+
+                Option<?> optionMechanismNames = Option.fromString("org.xnio.Options."+ Options.SASL_MECHANISMS.getName(), loader);
+                String[] mechanismNames = null;
+                if(connectorPropertiesOptionMap.contains(optionMechanismNames)) {
+                    Object o = connectorPropertiesOptionMap.get(optionMechanismNames);
+                    if (o instanceof Sequence) {
+                        Sequence<String> sequence = (Sequence<String>) connectorPropertiesOptionMap.get(optionMechanismNames);
+                        mechanismNames = sequence.toArray(new String[sequence.size()]);
+                    }
+                }
+
+                Option<?> optionPolicyNonanonymous = Option.fromString("org.xnio.Options."+ Options.SASL_POLICY_NOANONYMOUS.getName(), loader);
+                //in case that legacy sasl mechanisms are used, noanonymous default value is true
+                Boolean policyNonanonymous = mechanismNames == null ? null: true;
+                if(connectorPropertiesOptionMap.contains(optionPolicyNonanonymous)) {
+                    Object o = connectorPropertiesOptionMap.get(optionPolicyNonanonymous);
+                    if (o instanceof Boolean) {
+                        policyNonanonymous = (Boolean) o;
+                    }
+                }
+
+                if(mechanismNames != null || policyNonanonymous != null) {
+                    saslAuthenticationFactory = securityRealm.getSaslAuthenticationFactory(mechanismNames, policyNonanonymous);
```

Comment:
I think rather than expanding this particular API, this should probably done using one or more wrapping SaslServerFactory around the normal SaslAuthenticationFactory SaslServerFactory.  We don't want XNIO options adding mechanisms that were not allowed by the global configuration: if anything we only want to reduce the set IMO.

Reply:
> I was discussing it firstly with Darran L. and the result of the discussion was to add getSaslAuthenticcationFactory with more parameters to propagate XNIO options.  It seems to me, that wrapping factory could be used only for reducing mechanisms, but this case should allow also adding new mechanisms.

OK we can continue that discussion.  One factor to consider is that we should not extend APIs for legacy constructs.

Reply:
So in this case as @JiriOndrusek says we needed the legacy realm to add support for two legacy configuration options, firstly we needed to make sure anonymous would be authorized and then secondly we needed to add it to the list of mechanism advertised.  In the Remoting subsystem I also believe we only have the option to reduce the set of mechanisms not increase it.

In the Remoting subsystem we probably should make sure the properties are deprecated if they have not already been as we don't use them with pure Elytron capabilities.

But for the remainder this is a bug fix in legacy code to resurrect some legacy behaviour, having the majority of the fix in the legacy code seems more appropriate as when possible that entire block of legacy code will be removed whilst the Remoting subsystem will predominantly remain.

Reply:
The SecurityRealm API however is a legacy API, and the getSaslAuthenticationFactory is specifically to get an Elytron SASL authentication factory that emulates legacy behaviour.

Reply:
Ah okay, that makes it more clear.  As long as this isn't a problem for the realms code then I'm fine with it.